File: Templates/R/Clustering/KMeansClustering.R

dataset <- read.csv('filename')
summary(dataset) # provides a summary of the dataset

str(dataset) # gives us a structure of the dataset

names(dataset) # names of the column

#check for correlation 

#tocheck for two columns

cor(dataset&column1,dataset&column2,method = 'pearson') # other methods can be used

#between all the variables

library(ggpubr)
ggscatter(dataset,x= 'column 1',y = 'column 2',add = 'reg.line',
          conf.int = TRUE , cor.coef = TRUE , cor.method = 'pearson',
          xlab = 'Column 1',ylab = 'Column 2')

#Correltion diagram

library(corrplot)
x <- cor(dataset)
corrplot(x, type="upper", order="hclust")

#Correlation Matrix

corr.test(dataset)

#Correlation of one variable will every variable

library(corrr)
dataset %>% correlate() %>% focus(columnname)

#Do scaling if needed
#Convert into categorical variable if required

#Before creating the model, you might need to bin few values . You can use mapvalue function for that
library(plyr)
dataset$column <- mapvalues(dataset$column,from = c('A','B','C'),to = c('D','E','F')) # to map A to D, B to E,c to F

#using the elbow method to find the optimal value of clusters
set.seed(123)
wcss <- vector()

for(i in 1:10)
  wcss[i] <- sum(kmeans(X,i)$withinss) 

#plotting the elbow curve
plot(1:10,wcss,type = 'b',main = paste('Clusters'),xlab = 'Number of Clusters',ylab = 'WCSS')


#Model 
model <- kmeans(X, 5,iter.max = 300, nstart = 10) # nstart are the number of initial random centroids we want.
                                                  # iter.max aare the number of iterations we want

  

File: Templates/Python/Clustering/KMeansClustering.py
# -*- coding: utf-8 -*-
"""
Created on Sun Jan 27 02:13:10 2019

@author: Kenil Shah
"""

import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

#Importing Dataset
dataset = pd.read_csv("File Name")

#Gives us a summary of the dataset
dataset.describe()

#to give column names
dataset.columns = []

#Diving the dataset into independent and dependent variables
X = dataset.iloc[:,:].values
Y = dataset.iloc[:,:].values

# Column names
list(dataset)

#Correlation Matrix
corr_matrix = dataset.corr()
f, ax = plt.subplots(figsize=(16,10))
ax = sns.heatmap(corr_matrix)
ax.set_title("correlation between all features")
figure = ax.get_figure()    

#Do feature scaling if required.
#Convert into categorical variable if required

#Finding optimal number of clusters using elbow method
from sklearn.cluster import KMeans
wcss = []

for i in range(1,11):
    model = KMeans(n_clusters = i,init = 'k-means++',max_iter = 300, n_init = 10,random_state = 0) 
    #n_clusters is the number of clusters we want. init is the initial choice we want to make about centroids.
    #max_iter is the numbers of iterations you want . n_init is the number of times this algorithm will be run with different centroids
    model.fit(dataset)
    wcss = wcss.append(kmeans.inertia_)

#Plotting the Elbow graph
plt.plot(range(1,11),wcss)
plt.title('The Elbow Method')
plt.xlabel('Number Of CLusters')
plt.ylabel('The WCSS')
plt.show()

#The Final Kmeans clustering
finalmodel = KMeans(n_clusters = 5,init = 'k-means++',max_iter = 300, n_init = 10,random_state = 0) 
predictedclusters = finalmodel.fit_predict(X)


    
